### PR TITLE
Add support for illumos target

### PR DESCRIPTION
**Pull request by jasonbking**
This change adds support for an illumos os target to libc.  Similar to the BSDs, there is a large deal of overlap (given the common history), so the 'solaris' directory was renamed to 'solarish' (it's the closest thing to an official term to refer to things descending from Solaris as well as Solaris).  There were also a number of missing definitions (as well as a couple missing functions) that have proved necessary for building a number of rust programs on illumos or Solaris. 

Portions contributed by @papertigers .

### CODE REVIEW

**Comment by jasonbking** [comment]
The AppVeyor failure appears to be a network related issue, however the Travis CI is flagging rustfmt.  I can run the code through rustfmt and update the PR, though should 2015 or 2018 style be used? It may not make a difference between the two, but it would be good to know in general for future PRs.  I didn't see anything that discussed it for submitting PRs, but might have missed it.

**Comment by gnzlbg** [comment]
> however the Travis CI is flagging rustfmt. I can run the code through rustfmt and update the PR, though should 2015 or 2018 style be used?

Does `cargo fmt --all` using the latest nightly `rustfmt-preview` component "do the right thing"? I probably need to add this information to the readme.

**Comment by jasonbking** [comment]
Unless I'm misunderstanding the output, it appears that `cargo fmt --all` isn't even looking at src/unix/solarish/mod.rs:

```
cargo fmt --all -v
[test (2015)] "/root/ws/libc/libc-test/test/cmsg.rs"
[custom-build (2015)] "/root/ws/libc/build.rs"
[test (2015)] "/root/ws/libc/libc-test/test/linux_fcntl.rs"
[test (2015)] "/root/ws/libc/libc-test/test/main.rs"
[custom-build (2015)] "/root/ws/libc/libc-test/build.rs"
[lib (2015)] "/root/ws/libc/src/lib.rs"
rustfmt --edition 2015 /root/ws/libc/libc-test/test/cmsg.rs /root/ws/libc/build.rs /root/ws/libc/libc-test/test/linux_fcntl.rs /root/ws/libc/libc-test/test/main.rs /root/ws/libc/libc-test/build.rs /root/ws/libc/src/lib.rs
```

Running `rustfmt --edition 2015 --check` on src/unix/solarish/mod.rs shows a number of changes, including to code that wasn't touched by this change.  Running the same thing on a src/unix/solaris/mod.rs from the latest (as of a few minutes ago) copy of master shows similar proposed changes (i.e. rustfmt is flagging changes on the current version in master).  I can go ahead and push and update with the fixed formatting if that's what's needed.

**Comment by gnzlbg** [comment]
What's your `rustfmt` version ?

**Comment by jasonbking** [comment]
rustfmt 1.0.1-dev (be135599 2018-12-10)

**Comment by gnzlbg** [comment]
So I can reproduce this, I've filled an issue upstream: https://github.com/rust-lang/rustfmt/issues/3341

Would you mind adding a commit that comments out the rustfmt build bot from the `.travis.yml` file ? I have no idea what's going on, and we can always re-enable that later.

File: .travis.yml [hunk 1]
```diff
@@ -22,12 +22,13 @@ matrix:
...
-      script:
-        - rustc ci/style.rs && ./style src
-        - cargo fmt --all -- --check
-      stage: tools-and-build-and-tier1
+    # Disabled due to rust-lang/rustfmt#3341
+    #- name: "Style"
+    #  install: rustup component add rustfmt-preview
+    #  script:
+    #    - rustc ci/style.rs && ./style src
+    #    - cargo fmt --all -- --check
```
Comment (gnzlbg) [inline .travis.yml]: Would it be possible to disable only this line? We'd like to keep using the other style checker, I forgot about it :D

Comment (jasonbking) [reply .travis.yml]: Just the cargo fmt line?

Comment (gnzlbg) [reply .travis.yml]: Yes

**Comment by gnzlbg** [comment]
Thank you!

@bors: r+

**Comment by gnzlbg** [comment]
Looks like style is failing: https://travis-ci.com/rust-lang/libc/jobs/176963291#L214

**Comment by gnzlbg** [comment]
Thank you!

**Comment by jasonbking** [comment]
With version 0.2.49 tagged, what is required to get the version published to crates.io?  It appears that 0.2.48 is still the latest published.

**Comment by gnzlbg** [comment]
I just needed to do a `cargo publish`. 0.2.49, feel free to ping me about a release the next time you need one, and thank you for working on this!

**Pull request merged**